### ISSUE TITLE
Possible bug in do_acquire?

### ISSUE THREAD

**Issue by pchico83** [andymccurdy/redis-py#554]
This is the current code for Lock.do_acquire:

```
    def do_acquire(self, token):
        if self.redis.setnx(self.name, token):
            if self.timeout:
                # convert to milliseconds
                timeout = int(self.timeout * 1000)
                self.redis.pexpire(self.name, timeout)
            return True
        return False
```

I think it has a bug since the process might die after `self.redis.setnx(self.name, token)` but before `self.redis.pexpire(self.name, timeout)`. What would happen in that case? Would it be better to use `self.redis.setex` with try/catch?

**Comment by andymccurdy** [comment]
I agree that using `SET` with the `nx` and `px` parameters would be optimal, but it will break anyone connecting to a Redis server version < 2.6.12. This is largely why the Python-based Lock version if still maintained as the LuaLock would also exclude users below 2.6.0.

We can likely add a 3rd Lock implementation based on the code that @23doors suggests above and do the same version checking we do for the LuaLock, preferring the new version, falling back to LuaLock if the server is running 2.6.0 - 2.6.11 and finally the old lock class for anyone using an older Redis server.

**Comment by wojcikstefan** [comment]
Hi there! Was this issue solved by #569 or is there still something that needs to be done before this is closed?

**Comment by andymccurdy** [comment]
This issue was resolved long ago. Sorry for keeping it open.
